---
title: "Integrate DC datasets"
author: "Massimo Andreatta & Santiago Carmona"
#output: html_notebook
---

```{r}
renv::restore()
```

Libraries
```{r}
#remotes::install_github('carmonalab/UCell')
#remotes::install_github('carmonalab/STACAS')
#remotes::install_github('carmonalab/ProjecTILs')

library(ggplot2)
library(Seurat)
library(tidyr)
library(SeuratObject)
library(Matrix)
library(patchwork)
library(remotes)
library(UCell)
library(STACAS)
```

#NB: for conversion between Scanpy format to matrix files, see export_data.py script

Read in data as Seurat objects (see dc_analysis.Rmd for data conversion)
```{r}
pdir <- "~/Dropbox/CSI/Datasets/Gerhard_JEM"

seurat.list <- readRDS(sprintf("%s/Gerhard_metanalysis_seuratlist.DC.anno.rds",pdir))

#Exclude Brown dataset
seurat.list <- seurat.list[names(seurat.list) != "brownSpleen"]

#only tumor samples
#seurat.list <- seurat.list[(c("qian","qianbc","qiancrc","qianlung","qianovarian","maier","zhang","zhangcrc","zilionis"))]
```

Apply simple T cell signature to identify doublets
```{r}
seurat.list <- lapply(seurat.list, function(x) {
  
  sig <- list("Tcell"=c("CD3D","CD3E","CD3G","TRAC","LCK","CD2"))
  x <- AddModuleScore_UCell(x, features = sig)
  x <- SmoothKNN(x, signature.names = "Tcell_UCell", k = 10)
  
  x$isdoublet <- FALSE
  x$isdoublet[x$Tcell_UCell_kNN>0.2] <- TRUE
  
  x
})
```
```{r}
lapply(names(seurat.list), function(n) {
  x <- seurat.list[[n]]
  a <- FeaturePlot(x, features=c("Tcell_UCell_kNN"))
  b <- DimPlot(x, group.by = "isdoublet") + ggtitle(n)
  a | b
})  
```

Remove T cell doublets
```{r}
seurat.list <- lapply(seurat.list, function(x) {
  subset(x, subset=`isdoublet`==FALSE)
})  
```


See batch effects
```{r fig.height=3}
merged <- Reduce(merge, seurat.list)

merged <- FindVariableFeatures(merged, nfeatures = 1000)
merged <- merged %>% ScaleData() %>% RunPCA(npcs=30) %>% RunUMAP(dims=1:30)

library(RColorBrewer)
palette <- brewer.pal(n=5, name="Set3")
names(palette) <- c("cDC1","cDC2","DC3","pDC","Multi")

#palette <- brewer.pal(n=6, name="Set3")
#names(palette) <- c("cDC1","cDC2","DC3","pDC","MoDC","Multi")


p1 <- DimPlot(merged, reduction="umap", group.by="Study") + theme(aspect.ratio = 1)
p2 <- DimPlot(merged, reduction="umap", group.by="scGate_multi", cols=palette) + theme(aspect.ratio = 1)
p1 | p2

ggsave("plots/Gerhard_sets_preintegration.pdf", height = 5, width=10)
```

Find anchors
```{r}
nfeatures=2000
ndim=40
set.seed(1234)
stacas_anchors <- FindAnchors.STACAS(seurat.list, 
                                     anchor.features = nfeatures,
                                     dims = 1:ndim)
```

```{r}
tree <- SampleTree.STACAS(stacas_anchors,
                          obj.names = names(seurat.list),
                          hclust.method = "ward.D")
```

Integrate data
```{r}
object_integrated <- IntegrateData.STACAS(stacas_anchors,
                                          sample.tree=tree,
                                              dims=1:ndim)
```

Dimred
```{r}
set.seed(123)
object_integrated <- object_integrated %>% ScaleData() %>% RunPCA(npcs=ndim) %>% RunUMAP(dims=1:ndim)
```

```{r fig.height=3}
library(RColorBrewer)
palette <- brewer.pal(n=5, name="Set3")
names(palette) <- c("cDC1","cDC2","DC3","pDC","Multi")

#palette <- brewer.pal(n=6, name="Set3")
#names(palette) <- c("cDC1","cDC2","DC3","pDC","MoDC","Multi")


p1 <- DimPlot(object_integrated, reduction="umap", group.by="Study") + theme(aspect.ratio = 1)
p2 <- DimPlot(object_integrated, reduction="umap", group.by="scGate_multi", cols=palette) + theme(aspect.ratio = 1)
p1 | p2

ggsave("plots/Gerhard_sets_STACAS_unsup.pdf", height = 5, width=10)
```


Unsupervised clustering
```{r}
k.param <- 20
res <- 0.2
set.seed(123)

object_integrated <- FindNeighbors(object_integrated, k.param=k.param, dims=1:ndim)
object_integrated <- FindClusters(object_integrated, resolution = res, graph.name = "integrated_snn")

DimPlot(object_integrated, reduction = "umap", group.by="seurat_clusters") + theme(aspect.ratio = 1)
```

Find markers for unsupervised clusters
```{r}
Idents(object_integrated) <- "seurat_clusters"
DefaultAssay(object_integrated) <- "RNA"

set.seed(1234)
cluster.markers <- FindAllMarkers(object_integrated, only.pos = T, min.pct = 0.1, min.diff.pct=0.1, 
                                  logfc.threshold = 0.25, max.cells.per.ident = 500, test.use="wilcox",base=exp(1))

all <- cluster.markers |> dplyr::group_by(cluster) |> dplyr::top_n(n = 50, wt = abs(avg_logFC))

for (i in levels(object_integrated@active.ident)) {
    print(subset(all, cluster==i))
}  
DefaultAssay(object_integrated) <- "integrated"

```
Pairwise differences
```{r}
pair.m <- FindMarkers(object_integrated, ident.1 = "1", ident.2 = "6", only.pos = TRUE)
head(pair.m, 20)

pair.n <- FindMarkers(object_integrated, ident.1 = "6", ident.2 = "1", only.pos = TRUE)
head(pair.n, 20)
```

Where are specific subtypes, e.g. the "DC5" by Villani et al.?
```{r fig.height=10, fig.align=16}
p <- list()
anns <- unique(object_integrated$state)

for (a in anns) {
p[[a]] <- DimPlot(object_integrated, reduction = "umap", cells.highlight = which(object_integrated$state == a)) + 
  theme(aspect.ratio = 1) + NoLegend() + ggtitle(a)
}

wrap_plots(p)
```

```{r fig.height=3}
DefaultAssay(object_integrated) <- "RNA"

FeaturePlot(object_integrated, features=c("CLEC10A","IL1R2","CD1C","CFP"))
FeaturePlot(object_integrated, features=c("C1QA","C1QB","C1QC","FCGR2A"))
FeaturePlot(object_integrated, features=c("CD1A","CD207","FCGBP","S100B"))
FeaturePlot(object_integrated, features=c("S100A8","S100A9","FCN1","CD14"))
FeaturePlot(object_integrated, features=c("S100B","ACY3","PRDM16","CD14"))
FeaturePlot(object_integrated, features=c("LTB","RUNX3","LTB","IL22RA2"))
FeaturePlot(object_integrated, features=c("PPP1R14A","DAB2","AXL","LILRA4"))
FeaturePlot(object_integrated, features=c("TRAC","CD2","CD3E","CD3D"))
FeaturePlot(object_integrated, features=c("TOP2A","STMN1","MKI67","CD3D"))

```

Try to name clusters
```{r}
object_integrated$annotation <- NA
object_integrated$annotation[object_integrated$seurat_clusters %in% c("0","8")] <- "cDC2_CLEC10A"
object_integrated$annotation[object_integrated$seurat_clusters %in% c("2","9")] <- "cDC1"
object_integrated$annotation[object_integrated$seurat_clusters %in% c("3")] <- "DC3"
object_integrated$annotation[object_integrated$seurat_clusters %in% c("1","6")] <- "cDC2_CD1A"
object_integrated$annotation[object_integrated$seurat_clusters %in% c("4")] <- "pDC"
object_integrated$annotation[object_integrated$seurat_clusters %in% c("5")] <- "MonoDC"
object_integrated$annotation[object_integrated$seurat_clusters %in% c("7")] <- "AS-DC"
object_integrated$annotation[object_integrated$seurat_clusters %in% c("10")] <- "PRDM16_DC"
DimPlot(object_integrated, group.by = "annotation")
```

Make it into reference map
```{r}
library(ProjecTILs)
DefaultAssay(object_integrated) <- 'integrated'
ref <- make.reference(object_integrated, recalculate.umap = TRUE,
                      annotation.column = "annotation", ndim = ndim)

```


```{r}
states <- levels(ref$functional.cluster)

palette <- brewer.pal(n = length(states), name = "Paired")
names(palette) <- states
ref@misc$atlas.palette <- palette

DimPlot(ref, group.by = "functional.cluster", label = TRUE, cols = palette) + theme(aspect.ratio = 1)
ggsave("plots/DC_reference_wTcells.pdf", height=5, width=7)
```

```{r fig.height=3.5}
DefaultAssay(ref) <- "RNA"
FeaturePlot(ref, features=c("CD1A","CD207","ENTPD1","FCGBP"))
FeaturePlot(ref, features=c("ACY3","PRDM16","LAMP3","TRAC"), order=T)
FeaturePlot(ref, features=c("LAMP3","BIRC3","GZMB","CCR7"))
FeaturePlot(ref, features=c("FCER1A","CD1C","CLEC10A","FCGR2A"))
FeaturePlot(ref, features=c("ZBTB46","ITGAX","CLEC10A","FCGR2A"))
```

```{r fig.height=3}
library(RColorBrewer)
ccc <- brewer.pal(n=length(ss)+1, name="Set3")
names(ccc) <- c(names(ss), "Multi")

p1 <- DimPlot(ref, reduction="umap", group.by="Study") + theme(aspect.ratio = 1)
p2 <- DimPlot(ref, reduction="umap", group.by="scGate_multi", cols=ccc) + theme(aspect.ratio = 1)
p1 | p2

ggsave("plots/DC_reference_bysample.pdf", height = 5, width=10)

#Save object
saveRDS(ref, "aux/DC_reference_wTcells.rds")
```

# Try dataset projection
```{r}
pdir <- "~/Dropbox/CSI/Datasets/Gerhard_JEM"

dataset.list <- readRDS(sprintf("%s/Gerhard_metanalysis_seuratlist.DC.anno.rds",pdir))

brown <- dataset.list$brownSpleen

DimPlot(brown, group.by = "state")
DimPlot(brown, group.by = "scGate_multi")
```

```{r}
brown.projected <- Run.ProjecTILs(brown, ref=ref, filter.cells = FALSE)
```

See results
```{r}
plot.projection(ref, brown.projected)
```

```{r fig.height=3, fig.width=8}
celltypes <- unique(brown.projected$state)
pll <- list()

for (c in celltypes) {
  sub <- subset(brown.projected, subset=state==c)
  pll[[c]] <- plot.projection(ref, query=sub, linesize = 0.5, pointsize = 0.5) + ggtitle(c)
}

wrap_plots(pll, ncol=4)

ggsave("plots/Brown_projected_bysubtype.png", height=9, width=22)
```

Composition
```{r}
plot.statepred.composition(ref=ref, query=brown.projected)
```

Profile agreement
```{r fig.height=4}
genes4radar <- c("CLEC9A","XCR1","CADM1","CLEC10A","CD1C","CD1A",
                 "LAMP3","CCR7","FSCN1","S100A8","S100A9","VCAN","FCN1",
                 "GZMB","LILRA4","TCF4","PPP1R14A","AXL","PRDM16","ACY3")

plot.states.radar(ref, query=brown.projected, genes4radar = genes4radar)
```

Where are specific subtypes, e.g. the "DC5" by Villani et al.?
```{r fig.height=9, fig.align=12}
p <- list()
anns <- unique(ref$state)

for (a in anns) {
p[[a]] <- DimPlot(ref, reduction = "umap", cells.highlight = which(ref$state == a)) + 
  theme(aspect.ratio = 1,
        axis.title=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank()) +
  NoLegend() + ggtitle(a)
}

wrap_plots(p)
ggsave("plots/DC_reference_byorig_annot.png", height=18, width=24)
```


```{r fig.height=4, fig.align=9, eval=F}
studies <- unique(ref$Study)
for (s in studies) {
  p <- list()
  this <- subset(ref, subset=Study==s)
  anns <- unique(this$state)
  for (a in anns) {
    p[[a]] <- DimPlot(ref, reduction = "umap", cells.highlight = which(ref$state == a)) + 
      theme(aspect.ratio = 1) + NoLegend() + ggtitle(a)
  }
  
  wrap <- wrap_plots(p)
  plot(wrap)
  ggsave(sprintf("plots/DC_reference_byorig_annot_split_%s.png",s), plot=wrap, height=5, width=13)
}
```

Find markers for subtypes in the reference
```{r}
Idents(ref) <- "functional.cluster"
DefaultAssay(ref) <- "RNA"

set.seed(1234)
cluster.markers <- FindAllMarkers(ref, only.pos = T, min.pct = 0.1, min.diff.pct=0.1, 
                                  logfc.threshold = 0.25, max.cells.per.ident = 500, test.use="wilcox",base=exp(1))

all <- cluster.markers |> dplyr::group_by(cluster) |> dplyr::top_n(n = 50, wt = abs(avg_logFC))

for (i in levels(ref@active.ident)) {
    print(subset(all, cluster==i))
}  
DefaultAssay(ref) <- "integrated"

```

Make a heatmap of DEGs
```{r fig.height=4}
ref.copy <- ref
DefaultAssay(ref.copy) <- 'RNA'
ref.copy <- ScaleData(ref.copy)

genes.use <- c()
for (i in levels(ref.copy$functional.cluster)) {
    this <- as.vector(all[all$cluster==i, "gene"])[[1]]
    this <- this[1:min(8, length(this))]
    
    genes.use <- c(genes.use, this)
}

heat <- DoHeatmap(ref.copy, assay="RNA", features=genes.use, group.by = "functional.cluster", group.colors = palette)

heat + scale_fill_gradientn(colors = c("blue", "white", "red"))
ggsave("plots/DC_scHeatmap_topDEG.png", height=11, width=21)
```


```{r fig.height=3}
DefaultAssay(ref.copy) <- "RNA"

data("genes.blocklist")
c1 <- genes.blocklist$Hs$cellCycle.G1S
c2 <- genes.blocklist$Hs$cellCycle.G2M
signatures <- list("Tcell"=c("CD2","CD3D","CD3E","CD3G","TRAC"), "cycling"=c(c1,c2))
ref.copy <- AddModuleScore_UCell(ref.copy, features=signatures, name = NULL)
ref.copy <- SmoothKNN(ref.copy, signature.names = c("Tcell","cycling"))
FeaturePlot(ref.copy, features=c("TRAC","CD3D","Tcell_kNN","MKI67","STMN1","cycling_kNN"), ncol=3)
ggsave("plots/DC_ref_Tcell_and_cycling.png", height=6, width=11)

DefaultAssay(ref.copy) <- "integrated"
```


Violin plot of marker genes
```{r fig.height=4, fig.width=3}
DefaultAssay(ref.copy) <- "RNA"

split <- SplitObject(ref.copy, split.by = "Study")

genes <- c("PPP1R14A","AXL","CLEC9A","XCR1","CADM1","CD1A","CD207","CLEC10A","CD1C",
                 "LAMP3","CCR7","FSCN1","S100A8","S100A9","VCAN","FCN1",
                 "GZMB","LILRA4","TCF4","PRDM16","ACY3","ITGAX","TRAC","CD3E","CD3D","CD3G")

VlnPlot(ref.copy, features = genes, stack = TRUE, cols = palette, flip = TRUE, fill.by = "ident")

ggsave("plots/DC_annotated_violins.pdf", height=8, width=6)

VlnPlot(split$maier, features = genes, stack = TRUE, cols = palette, flip = TRUE, fill.by = "ident")
VlnPlot(split$qian, features = genes, stack = TRUE, cols = palette, flip = TRUE, fill.by = "ident")
VlnPlot(split$zilionis, features = genes, stack = TRUE, cols = palette, flip = TRUE, fill.by = "ident")
VlnPlot(split$qianlung, features = genes, stack = TRUE, cols = palette, flip = TRUE, fill.by = "ident")
VlnPlot(split$qiancrc, features = genes, stack = TRUE, cols = palette, flip = TRUE, fill.by = "ident")
VlnPlot(split$martin, features = genes, stack = TRUE, cols = palette, flip = TRUE, fill.by = "ident")

```

Subset composition
```{r}
tab <- table(ref$Study, ref$functional.cluster)
tab.norm <- tab/rowSums(tab)

df <- reshape2::melt(tab.norm)
colnames(df) <- c("Study","Subtype","Frequency")

palette <- ref@misc$atlas.palette

ggplot(df, aes(x = Study, y = Frequency, 
            fill = Subtype)) + geom_bar(stat = "identity") + 
            theme_bw() + scale_fill_manual(values = palette) +
            theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
ggsave("plots/DC_composition_by_dataset.pdf", height=4, width=5)
```

Make a subset vs. study heatmap of pseudobulk profiles
```{r}
make.heatmap <- function(data, assay="RNA", genes, scale="row",
                         method=c("ward.D", "average"), brewer.palette="RdBu",
                         cluster.col = "functional.cluster", flip=FALSE,
                         cluster_genes = FALSE, cluster_samples=FALSE, min.cells = 10,
                         palette = NULL) {
  
  require(pheatmap)
  set.seed(123)
  
  method = method[1]
  
  data$Study_Subset <- factor(paste(data@meta.data[,cluster.col],data$Study,sep="."))
  
  t <- table(data$Study_Subset)
  accept <- names(t)[t>min.cells]
  
  data <- subset(data, subset=Study_Subset %in% accept)
   
  #Calculate mean expression by cluster
  m <- c()
  for( g in unique(genes)){
    m[[g]] <- tapply(data@assays[[assay]][g,],data$Study_Subset, mean)
  }
  m <- as.data.frame(m)
  
  m <- m[accept,]
  
  m.study <- factor(unlist(lapply(strsplit(rownames(m),"\\.",perl = T),function(x) x[[2]])))
  m.subset <- factor(unlist(lapply(strsplit(rownames(m),"\\.",perl = T),function(x) x[[1]])))

  #gaps_col <- (which(!duplicated(m.subset))-1)[-1]
  
  breaksList = seq(-2, 2, by = 0.1)
  
  require(RColorBrewer)
  color = colorRampPalette(rev(brewer.pal(n = 7, name = brewer.palette)))(length(breaksList))
  
  if (is.null(palette)) {
    palette = brewer.pal(n=length(unique(m.subset)), name="Paired")
    names(palette) <- unique(m.subset)
  }
  
  annotation_col = data.frame(
                    Subtype = m.subset,
                    Study = m.study
                )
  rownames(annotation_col) = rownames(m)


  if (flip) { 
    h <- pheatmap::pheatmap(m, cluster_rows = cluster_samples,
                            cluster_cols = cluster_genes,scale = scale,
                            breaks = breaksList, color=color, 
                            annotation_row = annotation_col, 
                            show_rownames = F,
                            border_color = NA,
                            annotation_colors = list(Subtype=palette), 
                            fontsize_row=6,fontsize = 7, 
                            clustering_method=method)
  } else {
    h <- pheatmap::pheatmap(t(m),cluster_rows = cluster_genes,
                            cluster_cols = cluster_samples,scale = scale,
                            breaks = breaksList, color=color, 
                            annotation_col = annotation_col, 
                            show_colnames = F,
                            border_color = NA,
                            annotation_colors = list(Subtype=palette), 
                            fontsize_row=6,fontsize = 7, 
                            clustering_method=method)
  }
  return(h)
}
```

Plot heatmap of pseudo-bulk by subtype/study
```{r}
library(pheatmap)
h <- make.heatmap(ref.copy, assay = "RNA", genes = genes.use, min.cells=15,
            palette=palette, cluster_genes = FALSE, cluster_samples = FALSE)

ggsave("plots/DC_heatmap_pseudobulk_unclustered.pdf", plot=h, height=6, width=9)

h <- make.heatmap(ref.copy, assay = "RNA", genes = genes.use, min.cells=20,
            palette=palette, cluster_genes = TRUE, cluster_samples = TRUE)
ggsave("plots/DC_heatmap_pseudobulk_clustered.pdf", plot=h, height=7, width=10)
```

# Simplify reference using SuperCell
```{r}
remotes::install_github("GfellerLab/SuperCell")
library(SuperCell)

gamma <- 20 # Graining level
npcs <- 40 

DefaultAssay(ref.copy) <- "integrated"
# Compute metacells using SuperCell package
MC <- SCimplify(
  X = GetAssayData(ref.copy), # single-cell log-normalized gene expression data
  genes.use = VariableFeatures(ref.copy), 
  gamma = gamma,
  n.pc = npcs
)

MC.ge <- supercell_GE(
  ge = GetAssayData(ref.copy),
  groups = MC$membership
)

MC.ge.RNA <- supercell_GE(
  ge = GetAssayData(ref.copy, assay="RNA"),
  groups = MC$membership
)


# Annotate metacells to functional cluster
MC$functional.cluster <- supercell_assign(
  cluster = ref.copy$functional.cluster,          # single-cell assignment to cell lines 
  supercell_membership = MC$membership,  # single-cell assignment to metacells
  method = "absolute" # available methods are c("jaccard", "relative", "absolute"), function's help() for explanation
)

# Compute purity of metacells
MC$purity <- supercell_purity(
  clusters = ref.copy$functional.cluster,
  supercell_membership = MC$membership, 
  method = "max_proportion"
)

# Metacell purity distribution
hist(MC$purity)

supercell_plot(
  MC$graph.supercells, 
  group = MC$functional.cluster, 
  seed = 1, 
  alpha = -pi/2,
  main  = "Metacells colored by subtype"
)

#Make a Seurat object
ref.coarse <- supercell_2_Seurat(
  SC.GE = MC.ge, 
  SC = MC, 
  fields = c("functional.cluster", "purity"),
  var.genes = MC$genes.use,
  N.comp = npcs
)

colnames(MC.ge.RNA) <- colnames(ref.coarse@assays$RNA@data)
ref.coarse[["uncorrected"]] <- CreateAssayObject(data = MC.ge.RNA)
ref.coarse <- ScaleData(ref.coarse, assay="uncorrected")

Idents(ref.coarse) <- "functional.cluster"

ref.coarse <- RunUMAP(ref.coarse, dims = 1:npcs)
DimPlot(ref.coarse)

```
Heatmap of gene expression based on MetaCells
```{r}
heat2 <- DoHeatmap(ref.coarse, assay="uncorrected", features=genes.use, group.by = "functional.cluster")

heat2 + scale_fill_gradientn(colors = c("blue", "white", "red"))

ggsave("plots/DC_scHeatmap_topDEG.metacells.png", height=11, width=16)
```

